elasticsearch_service/helpers/consumer.go

```diff
@@ -0,0 +1,41 @@
+package helpers
+
+import (
+	"context"
+	"fmt"
+	"github.com/segmentio/kafka-go"
+	"log"
+	"strings"
+)
+
+type KafkaConsumer struct {
+	Reader   *kafka.Reader
+	EsClient ElasticsearchClient
+}
+
+func (consumer *KafkaConsumer) Consume() {
+	//go func() {
+	fmt.Println("KafkaConsumer is ready to consume new messages.")
+	for {
+
+		m, err := consumer.Reader.ReadMessage(context.Background())
```

Comment:
If you look inside this method, you can see that it first calls FetchMessage and the CommitMessages. In our case here, we need to use these two methods separately like:
1. FetchMessage
2. Use this message for some business logic, in our case,  prepare it end send to elasticsearch endpoint
3. CommitMessages
The main reason for this is for us to avoid data loss and/or corruption. This Project now uses several different APIs which are being coordinated to perform various actions. You should treat any pipeline action ( like this one ) as a transaction. What I suggest is to incorporate a retry mechanism, which would perform a backoff until the Kafka index is being successfully committed.

Reply:
The offset is no longer commited before the message is consumed properly.